patches/tModLoader/Terraria/ModLoader/Default/UnloadedPosIndexing.cs

```diff
@@ -0,0 +1,70 @@
+using System.Collections.Generic;
+using Terraria.ModLoader.IO;
+
+//TODO: Refactor this and make static later when looking at locational NBT systems.
+
+namespace Terraria.ModLoader.Default
+{
+	internal class UnloadedPosIndexing
+	{
+		public int posID;
+
+		public UnloadedPosIndexing(int posX, int posY) {
+			// The order is determined in accordance with increasing Y in TileIO ReadModData such that PosID is ordered numerically
+			posID = posX * Main.maxTilesY + posY;
+		}
+
+		public UnloadedPosIndexing(int posID) {
+			this.posID = posID;
+		}
+
+		public void GetCoords(out int posX, out int posY) {
+			posY = posID % Main.maxTilesY;
+			posX = posID / Main.maxTilesY;
+		}
+
+		public void MapPosToInfo(List<UnloadedInfo> infos, List<TileIO.posMap> posMap, List<TileIO.posMap> prevPosMap = null, UnloadedInfo info = null)  {
+			ushort pendingID = 0;
+			if (info == null && prevPosMap.Count > 0) {
+				pendingID = (ushort) (FloorGetValue(prevPosMap.ToArray()) + infos.Count);
+			}
+			else {
+				pendingID = (ushort) infos.IndexOf(info);
+			}
+			
+			posMap.Add(new TileIO.posMap {
+				posID = this.posID,
+				infoID = pendingID
+			});
+		}
+
+		public ushort FloorGetValue(TileIO.posMap[] posMap) {
+			
+			int index = BinarySearchPosMap(posMap);	
+
+			return posMap[index].infoID;
+		}
+
+		private int BinarySearchPosMap (TileIO.posMap[] posMap) {
+			int minimum = 0, maximum = posMap.Length;
+			// Binary search for interval containing posID
+			while (maximum - minimum > 1) {
+				int split = (minimum + maximum) / 2;
+
+				if (split == posMap.Length - 1) {
+					break;
+				}
+
+				if (posMap[split].posID <= this.posID) {
+					minimum = split;
+				}
+
+				if (posMap[split + 1].posID > this.posID) {
+					maximum = split;
+				}
+
```

Comment:
Given we are looking for the interval from posMap[split] to posMap[split + 1] containing posID, it may be more effective to separate it out as it decreases the amount of iterations for elements closer to center. IIRC it should be net beneficial, as >80% of domain benefits.

An example being that if the posID is in the dead middle (say 50.1 / 100) - if I use if else, it would find the minimum is 50, but then take another few checks to determine the maximum is also 50.